controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionFalse, failureReason, "")
+		remediation.IncrementFailureCount(hr)
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityError,
+			fmt.Sprintf("Helm release failed: %s", err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	// Register the post release values
+	v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionTrue, successReason, "")
+	hr.Status.LastAppliedRevision = hr.Status.LastAttemptedRevision
+	hr.Status.LastReleaseRevision = util.ReleaseRevision(rel)
+	r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityInfo, fmt.Sprintf("Helm release succeeded"))
+
+	return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+}
+
+func (r *HelmReleaseReconciler) reconcileTest(ctx context.Context, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// Tests are not enabled for this release, return.
+	if !hr.Spec.GetTest().Enable {
+		return ctrl.Result{}, nil
+	}
+
+	// If we already run the test, return
+	if tested := meta.GetCondition(hr.Status.Conditions, v2.TestSuccessCondition); tested != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// If the release did not succeed, return
+	released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition)
+	if released == nil || released.Status == v1.ConditionFalse {
+		return ctrl.Result{}, nil
+	}
+
+	if _, err := run.Test(*hr); err != nil {
+		v2.SetHelmReleaseCondition(hr, v2.TestSuccessCondition, v1.ConditionFalse, v2.TestFailedReason, "")
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityError,
+			fmt.Sprintf("Helm test failed: %s", err.Error()))
+
+		remediation := hr.Spec.GetInstall().GetRemediation()
+		if released.Reason == v2.UpgradeSucceededReason {
+			remediation = hr.Spec.GetUpgrade().GetRemediation()
+		}
+
+		if remediation.MustIgnoreTestFailures(hr.Spec.GetTest().IgnoreFailures) {
+			return ctrl.Result{}, nil
+		}
+
+		remediation.IncrementFailureCount(hr)
+		v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionFalse, v2.TestFailedReason, "")
+		return ctrl.Result{}, err
+	}
+
+	v2.SetHelmReleaseCondition(hr, v2.TestSuccessCondition, v1.ConditionTrue, v2.TestSucceededReason, "")
+	r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityInfo, "Helm test passed")
+	return ctrl.Result{}, nil
+}
+
+func (r *HelmReleaseReconciler) reconcileRemediation(ctx context.Context, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// If the release did not explicitly fail, we should not remediate
+	released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition)
+	if released == nil || released.Status != v1.ConditionFalse {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the right remediation strategy based on the failure count,
+	// looking back, this would have been easier if we did not decide
+	// to flatten our conditions (https://github.com/fluxcd/helm-controller/pull/75)
```

Comment:
@seaneagan please have a look at this in the context of the code in this PR (please note, it is an early stage draft and some details are not 100% right yet, I plan to resolve those while working on tests in Go).

Reply:
What is not reflected here, is that I originally started with also splitting the `reconcileRelease` into a `reconcileInstall` and `reconcileUpgrade`. Not knowing exactly what happened during a previous reconciliation (even while taking into account the failure counts), made this impossible.

It is true that you would still need to cobble together conditions at times, but it would made state determination much easier, and the code much more readable. At current it feels like we are at times following tiny breadcrumbs to determine state, which is also a quite brittle approach in my opinion.

Reply:
@hiddeco The `Released` condition already exposes `InstallFailed` or `UpgradeFailed` reasons. Currently it also includes `TestFailed` in its determination, but if we simply remove keep the helm test result encapsulated in the `TestSuccess` condition only, and record the `Released` condition directly after the install/upgrade (i.e. in `reconcileRelease` / `reconcileInstall` / `reconcileUpgrade`), then this should provide the explicit previous install/upgrade failure detection we are looking for here?